File: queue.js
// Time Complexity : Worst case O(n) for pop, O(1) for rest
// Space Complexity : O(n) for push and O(1) for rest
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Accessing the first element in the stack for popping was a challenge. 


/**
 * Initialize your data structure here.
 */
var MyQueue = function() {
    this.stack1  = [];
    this.stack2 = [];
};

/**
 * Push element x to the back of queue. 
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function(x) {
    this.stack1.push(x);
};

//method to push all the elements in stack 1 to stack 2 in reverse order. So this makes it easy to pop and peek. 
MyQueue.prototype._pushBack = function() {
    while(this.stack1.length > 0) {
        this.stack2.push(this.stack1.pop());
    }
}

/**
 * Removes the element from in front of queue and returns that element.
 * @return {number}
 */

MyQueue.prototype.pop = function() {
    if(this.empty()) return null

    if(this.stack2.length === 0){
        this._pushBack()
    }
    return this.stack2.pop()
};

/**
 * Get the front element.
 * @return {number}
 */
MyQueue.prototype.peek = function() {
    if(this.stack1.length > 0){
        return this.stack1[0]
    }
    return this.stack2[this.stack2.length -1];
    
};

/**
 * Returns whether the queue is empty.
 * @return {boolean}
 */
MyQueue.prototype.empty = function() {
    return this.stack1.length === 0 && this.stack2.length === 0
};

/** 
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */

File: queue.cpp
//232. Implement Queue using Stacks
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// I have taken two stacks if i want the first element of the stack then, i will return the top element in out stack 
// if the out stack is empty then i will pop all the elements in the in stack and push to the out stack then get the top element
// pop opertion is also the same method as pee operation.
class MyQueue {
public:
    stack<int> in;
    stack<int> out;
    void push(int x) {
        in.push(x);
    }
    
    int pop() {
       if(out.empty())
        {
            while(!in.empty())
            {
                out.push(in.top());
                in.pop();
            }
        }
        int x = out.top();
        out.pop();
        return x;
    }
    
    int peek() {
        if(out.empty())
        {
            while(!in.empty())
            {
                out.push(in.top());
                in.pop();
            }
        }
        return out.top();
    }
    
    bool empty() {
        if(in.size() == 0 && out.size() == 0) return true;
        return false;
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */